File: 2.add-two-numbers.js
/*
 * @lc app=leetcode id=2 lang=javascript
 *
 * [2] Add Two Numbers
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) {
  let currNode = head = new ListNode()
  let carry = 0

  while (l1 || l2) {
    currNode.next = new ListNode()
    currNode = currNode.next
    let val = (l1 ? l1.val : 0) + (l2 ? l2.val : 0) + carry
    if (val >= 10) {
      currNode.val = val % 10
      carry = 1
    } else {
      currNode.val = val
      carry = 0
    }
    l1 = l1 ? l1.next : null
    l2 = l2 ? l2.next : null
  }
  if (carry !== 0) {
    currNode.next = new ListNode(carry)
  }
  return head.next
};
// @lc code=end

File: csharp/2.add-two-numbers.cs
/*
 * @lc app=leetcode id=2 lang=csharp
 *
 * [2] Add Two Numbers
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution
{
  public ListNode AddTwoNumbers(ListNode l1, ListNode l2)
  {
    var res = new ListNode();
    var curr = res;
    int carry = 0;
    while (l1 != null || l2 != null)
    {
      int val1 = 0;
      int val2 = 0;
      if (l2 == null)
      {
        val1 = l1.val;
      }
      else if (l1 == null)
      {
        val2 = l2.val;
      }
      else
      {
        val1 = l1.val;
        val2 = l2.val;
      }
      int sum = val1 + val2 + carry;
      carry = sum / 10;
      var newNode = new ListNode(sum % 10);
      curr.next = newNode;
      curr = curr.next;
      l1 = l1?.next;
      l2 = l2?.next;
    }
    if (carry > 0)
    {
      var newNode = new ListNode(carry);
      curr.next = newNode;
      curr = curr.next;
    }

    return res.next;
  }
}
// @lc code=end

